File: LeetcodeProblems/blind-75/dp/combination-sum/solution.py
# Combination Sum - LeetCode #39

from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        
        def backtrack(temp: List[int], remain: int, start: int):
            if remain < 0:
                return
            if remain == 0:
                result.append(temp[:])
                return
            
            for i in range(start, len(candidates)):
                temp.append(candidates[i])
                backtrack(temp, remain - candidates[i], i)
                temp.pop()
        
        backtrack([], target, 0)
        return result

File: LeetcodeProblems/blind-75/dp/combination-sum/solution.js
// Combination Sum - LeetCode #39

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const result = [];
    
    function backtrack(temp, remain, start) {
        if (remain < 0) return;
        if (remain === 0) {
            result.push([...temp]);
            return;
        }
        
        for (let i = start; i < candidates.length; i++) {
            temp.push(candidates[i]);
            backtrack(temp, remain - candidates[i], i);
            temp.pop();
        }
    }
    
    backtrack([], target, 0);
    return result;
};
